R/run.R

```diff
@@ -18,42 +36,72 @@
 #' run_tutorial("hello", "learnr")
 #' }
 #'
-#' @importFrom utils adist
 #' @export
-run_tutorial <- function(name = NULL, package = NULL, shiny_args = NULL) {
-
-  if (is.null(package) && !is.null(name)) {
-    stop.("`package` must be provided if `name` is provided.")
-  }
+run_tutorial <- function(
+  name = NULL,
+  package = NULL,
+  ...,
+  shiny_args = NULL,
+  clean = FALSE,
+  as_rstudio_job = NULL
+) {
+  ellipsis::check_dots_empty()
+  checkmate::assert_character(name, any.missing = FALSE, max.len = 1, null.ok = TRUE)
+  checkmate::assert_character(package, any.missing = FALSE, max.len = 1, null.ok = TRUE)
 
-  # works for package = NULL and if package is provided
-  tutorials <- available_tutorials(package = package)
   if (is.null(name)) {
+    tutorials <- available_tutorials(package = package)
     message(format(tutorials))
     return(invisible(tutorials))
   }
 
-  # get path to tutorial
-  tutorial_path <- get_tutorial_path(name, package)
+  # The tutorial object must have a boolean `valid` indicating if a valid
+  # tutorial was found, and `dir` the normalized absolute path to the directory
+  # containing the tutorial. Optionally, a `file` item will also be included,
+  # indicating which specific tutorial will be run.
+  tutorial <-
+    if (is.null(package)) {
+      run_validate_tutorial_path(name)
+    } else {
+      run_validate_tutorial_pkg(name, package)
+    }
+
+  if (!isTRUE(tutorial$valid)) {
+    run_stop_invalid_name(name = name, package = package)
+  }
 
   # check for necessary tutorial package dependencies
-  install_tutorial_dependencies(tutorial_path)
+  install_tutorial_dependencies(tutorial$dir)
 
   # provide launch_browser if it's not specified in the shiny_args
-  if (is.null(shiny_args))
+  if (is.null(shiny_args)) {
     shiny_args <- list()
+  }
   if (is.null(shiny_args$launch.browser)) {
-    shiny_args$launch.browser <- (
-      interactive() ||
-        identical(Sys.getenv("LEARNR_INTERACTIVE", "0"), "1")
-    )
+    is_interactive <- rlang::is_interactive() ||
+      identical(Sys.getenv("LEARNR_INTERACTIVE", "0"), "1")
+
+    shiny_args$launch.browser <- if (!is_interactive) {
+      utils::browseURL
+    } else {
+      getOption("viewer", utils::browseURL)
+    }
+  }
+
+  as_rstudio_job <-
+    (isTRUE(as_rstudio_job) && can_run_rstudio_job(TRUE)) ||
+    (is.null(as_rstudio_job) && can_run_rstudio_job())
```

Comment:
```suggestion
    (isTRUE(as_rstudio_job) && can_run_rstudio_job(stop_if_not = TRUE)) ||
    (is.null(as_rstudio_job) && can_run_rstudio_job())
```

Including the argument name makes it a little clearer why the two calls to `can_run_rstudio_job()` differ.

Reply:
Thanks! I stumbled on that line when returning to this PR so I agree it's a good idea to include that argument name.